src/lv_init.c

```diff
@@ -52,6 +55,83 @@ static bool lv_initialized = false;
 /**********************
  *   GLOBAL FUNCTIONS
  **********************/
+static inline void lv_global_init(lv_global_t * global)
+{
+    LV_ASSERT_NULL(global);
+
+    if(global == NULL) {
+        return;
+    }
+
+    memset(global, 0, sizeof(lv_global_t));
+
+    global->style_refresh = true;
+    global->layout_count = _LV_LAYOUT_LAST;
+    global->memory_zero = ZERO_MEM_SENTINEL;
+    global->style_last_custom_prop_id = (uint16_t)_LV_STYLE_LAST_BUILT_IN_PROP;
+    global->area_trans_cache.angle_prev = INT32_MIN;
+    global->event_last_register_id = _LV_EVENT_LAST;
+    global->math_rand_seed = 0x1234ABCD;
+
+#if defined(LV_DRAW_SW_SHADOW_CACHE_SIZE) && LV_DRAW_SW_SHADOW_CACHE_SIZE > 0
+    global->draw_cache.sw_shadow_cache.cache_size = -1;
+    global->draw_cache.sw_shadow_cache.cache_r = -1;
+#endif
+}
+
+#ifdef LV_THREAD_LOCAL
+
+lv_global_t * lv_global_default(void)
+{
+    static LV_THREAD_LOCAL lv_global_t lv_global;
+    return &lv_global;
+}
+
+#else
+
+static pthread_key_t lv_global_key;
+
+static inline lv_global_t * lv_global_create(void)
+{
+    lv_global_t * global = malloc(sizeof(lv_global_t));
+    LV_ASSERT_MALLOC(global);
+
+    lv_global_init(global);
+    return global;
+}
+
+static void lv_global_destroy(void * data)
+{
+    if(data) {
+        free(data);
+    }
+}
+
+static inline void lv_global_cleanup(void)
+{
+    (void) pthread_setspecific(lv_global_key, NULL);
+}
+
+static void lv_global_key_init(void)
+{
+    (void) pthread_key_create(&lv_global_key, lv_global_destroy);
+}
+
+lv_global_t * lv_global_default(void)
+{
+    static pthread_once_t key_once = PTHREAD_ONCE_INIT;
+    (void) pthread_once(&key_once, lv_global_key_init);
+    lv_global_t * data;
+
+    if((data = (lv_global_t *)pthread_getspecific(lv_global_key)) == NULL) {
+        data = lv_global_create();
+        LV_ASSERT_NULL(data);
+
+        (void) pthread_setspecific(lv_global_key, data);
+    }
+    return data;
+}
```

Comment:
We can move this part out of LVGL. When LV_THREAD_LOCAL is not defined, it means that the porting layer needs to implement the lv_global_default method to support multiple instances. 

If add lv_init_instance,  it is not clear how to implement the differences between multiple instances.

Reply:
From your example, I understand that the `malloc` and `free` functions of the `lv_global_t` structure do not necessarily have to be provided by LVGL, but the initialization of the structure is necessary. But besides initialization, how does `lv_init_instance` save the pointer to achieve binding with the current thread? 

Do you mean saving to a variable with the `__thread` keyword? AFAIK, in some systems, this keyword may not be supported. It may be necessary to use some platform-specific APIs or libraries to achieve thread-local storage. This means that the implementation of `lv_global_default` may vary depending on the system. we need to support custom implementations and cannot define a standard implementation approach in LVGL.

Reply:
I see, we really need `LV_THREAD_LOCAL`, but can this be enough?
```c
static LV_THREAD_LOCAL lv_global_t * lv_global;
void lv_init_instance(lv_global_t * g) //Normal pointer, not thread specific
{ 
   lv_global = g; //Save in LV_THREAD_LOCAL prefixed pointer
   lv_init();
}

void lv_init(void)
{
  if(lv_global == NULL) lv_global = lv_malloc(sizeof(lv_global_t));
  
   ...
}
```

What I don't really understand is the need for the pthread related part in Line 91-132 when LV_THREAD_LOCAL is not defined. Can you tell why is it required inside LVGL?

Reply:
The following description is based on the absence of `__thread `support.

In your example, a static `lv_global_t*` is defined, which behaves as follows in systems without the concept of processes. When thread1 is initialized, `lv_global `is assigned the value addr1, but when thread2 is initialized, `lv_global `is assigned the value addr2. At this point, when thread1 tries to get `lv_global`, it will return addr2 instead of addr1. This is not the expected behavior.

The codes related part in Line 91-132 implements the ability to obtain an `lv_global_t` instance that is associated with the current thread in different threads. We don't have to implement this code in LVGL. LVGL just needs to allow the use of an externally implemented `lv_global_default `method to get lv_global_t*.

for example:
```
//demo.c

lv_global_t * lv_global_default(void) {
...
}

int demo_entry()
{
    lv_global_t * g = lv_global_default();
    //Any magic with `g`
    lv_init_instance(g);

    ...

    lv_deinit();
    lv_free(g);
}
```

```
//lvgl
lv_global_t * lv_global_default(void);

void lv_init_instance(lv_global_t * g) 
{ 
   // init g

   lv_init();
}

```

Reply:
My assumption was that if we add the `LV_THREAD_LOCAL` prefix to `lv_global_t * lv_global` it will mean the address saved by its thread (`addr1` or `addr2` based on which thread asks it). If it's not the case we can really 
- implement a simple non threaded `lv_global_default()` is `LV_THREAD_LOCAL` is no defined or `0` 
- let the user implement  `lv_global_default()` is `LV_THREAD_LOCAL` defined.

If we are on the same page now, could you update the PR? :slightly_smiling_face:

Reply:
I suggest using `LV_GLOBAL_CUSTOM` to indicate whether the user wants to customize the `lv_global_default()` implementation. `LV_THREAD_LOCAL` is only used to identify the `lv_global` variable in the default `lv_global_default`.
- implement a simple `lv_global_default()` if `LV_GLOBAL_CUSTOM` is no defined.
- let the user implement `lv_global_default()` if `LV_GLOBAL_CUSTOM` defined.

If the user system supports `__thread` in this scenario, then `#define LV_THREAD_LOCAL __thread` allows for the direct use of the default implementation of `lv_global_default`.

What do you think? If possible, I will update the PR later.

Reply:
`LV_GLOBAL_CUSTOM` sounds good!

> If the user system supports __thread in this scenario, then #define LV_THREAD_LOCAL __thread allows for the direct use of the default implementation of lv_global_default.

If `__thread` is not required internally, I'd leave it to the user and mention some use cases options in the documentation instead. Actually I think it'd be enough to document these two use cases with `__thread` and `pthread_key`. If I understand correctly these are independent solutions. If `__thread` is supported, it's great as it's simpler, but if not there is a workaround using `pthread_key`. Can you confirm it?

Reply:
Yes. So you mean you prefer to leave both options for the user to implement, right? LVGL only keeps the following implementation internally.

Reply:
Yes, but to be more precise:
```
#ifndef LV_GLOBAL_CUSTOM
lv_global_t * lv_global_default(void)
{
    static lv_global_t lv_global;
    return &lv_global;
}
#endif
```